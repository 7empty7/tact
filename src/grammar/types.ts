//
// Values
//

export type GrammarNumber = {
    kind: 'number',
    value: bigint
}

export type GrammarID = {
    kind: 'id',
    value: string
}

export type GrammarNull = {
    kind: 'null'
}

export type GrammarBoolean = {
    kind: 'boolean',
    value: boolean
}

//
// Expressions
//

export type GrammarOpBinary = {
    kind: 'op_binary',
    op: '+' | '-' | '*' | '/',
    left: GrammarExpression,
    right: GrammarExpression,
}

export type GrammarOpUnary = {
    kind: 'op_unary',
    op: '+' | '-' | '!' | '~',
    right: GrammarExpression,
}

export type GrammarOpField = {
    kind: 'op_field'
    src: GrammarExpression,
    key: string
}

export type GrammarOpCall = {
    kind: 'op_call'
    src: GrammarExpression,
    key: string,
    args: GrammarExpression[]
}

//
// Program
//

export type GrammarProgram = {
    kind: 'program',
    entries: (GrammarStruct | GrammarContract)[]
}

export type GrammarStruct = {
    kind: 'def_struct',
    name: string
    fields: GrammarField[]
}

export type GrammarField = {
    kind: 'def_field',
    name: string,
    type: string
}

export type GrammarContract = {
    kind: 'def_contract',
    name: string,
    declarations: (GrammarField)[]
}

export type GrammarArgument = {
    kind: 'def_argument',
    name: string,
    type: string
}

export type GrammarFunction = {
    kind: 'def_function',
    name: string,
    return: string,
    args: GrammarArgument[],
    statements: GrammarStatement[]
}

//
// Statements
//

export type GrammarStatementLet = {
    kind: 'let',
    name: string,
    expression: GrammarExpression
}

export type GrammarStatementReturn = {
    kind: 'return',
    expression: GrammarExpression
}

//
// Unions
//

export type GrammarStatement = GrammarStatementLet | GrammarStatementReturn;
export type GrammarExpression = GrammarOpBinary | GrammarOpUnary | GrammarOpField | GrammarNumber | GrammarNull | GrammarID | GrammarBoolean;
export type GrammarNodes = GrammarExpression | GrammarProgram | GrammarStruct | GrammarField | GrammarContract | GrammarArgument | GrammarFunction | GrammarOpCall | GrammarStatementLet | GrammarStatementReturn;