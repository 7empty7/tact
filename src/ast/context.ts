import { parse } from "../grammar/grammar";
import { ASTNode, ASTType } from "./ast";

export type VariableRef = {
    name: string,
    node: ASTNode
};

export class CompilerContext {

    static fromSources(sources: string[]) {
        let asts = sources.map(source => parse(source));
        let ctx = new CompilerContext();
        for (let a of asts) {
            for (let e of a.entries) {
                if (e.kind === 'def_struct') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'def_contract') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'primitive') {
                    ctx = ctx.addASTType(e);
                }
            }
        }
        return ctx;
    }

    readonly astTypes: { [key: string]: ASTType };
    readonly variables: { [key: string]: VariableRef };
    readonly shared: { [key: symbol]: any } = {};

    constructor(astTypes: { [key: string]: ASTType } = {}, variables: { [key: string]: VariableRef } = {}, shared: { [key: symbol]: any } = {}) {
        this.astTypes = astTypes;
        this.variables = variables;
        this.shared = shared;
        Object.freeze(this.astTypes);
        Object.freeze(this.shared);
        Object.freeze(this.variables);
        Object.freeze(this);
    }

    addASTType = (ref: ASTType) => {
        if (this.astTypes[ref.name]) {
            throw Error('Type already exists');
        }
        return new CompilerContext({ ...this.astTypes, [ref.name]: ref }, this.variables, this.shared);
    }

    addVariable = (ref: VariableRef) => {
        if (this.variables[ref.name]) {
            throw Error('Variable already exists');
        }
        return new CompilerContext(this.astTypes, { ...this.variables, [ref.name]: ref }, this.shared);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext(this.astTypes, this.variables, { ...this.shared, [store]: sh });
    }
}

export function createContextStore<T>() {
    let symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            let m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}