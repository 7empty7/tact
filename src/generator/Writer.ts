import { enabledDebug } from "../config";
import { CompilerContext } from "../context";
import { topologicalSort } from "../utils/utils";
import { Writer } from "../utils/Writer";

export class WriterContext {

    readonly ctx: CompilerContext;
    #functions: Map<string, { name: string, code: string, depends: Set<string> }> = new Map();
    #pendingWriter: Writer | null = null;
    #pendingDepends: Set<string> | null = null;
    #pendingName: string | null = null;
    #nextId = 0;
    #headers: string[] = [];

    constructor(ctx: CompilerContext) {
        this.ctx = ctx;
    }

    allocateNextRandomName() {
        return `__gen_internal_${this.#nextId++}`;
    }

    clone() {
        let res = new WriterContext(this.ctx);
        res.#functions = new Map(this.#functions);
        res.#nextId = this.#nextId;
        res.#headers = [...this.#headers];
        return res;
    }

    //
    // Rendering
    //

    render(debug: boolean = false) {

        // Check dependencies
        let missing = new Map<string, string[]>();
        for (let f of this.#functions.values()) {
            for (let d of f.depends) {
                if (!this.#functions.has(d)) {
                    if (!missing.has(d)) {
                        missing.set(d, [f.name]);
                    } else {
                        missing.set(d, [...missing.get(d)!!, f.name]);
                    }

                }
            }
        }
        if (missing.size > 0) {
            throw new Error(`Functions ${Array.from(missing.keys()).map((v) => `"${v}"`).join(', ')} wasn't rendered`);
        }

        // All functions
        let all = Array.from(this.#functions.values());

        // Remove unused
        if (!debug) {
            let used = new Set<string>();
            let visit = (name: string) => {
                used.add(name);
                let f = this.#functions.get(name)!!;
                for (let d of f.depends) {
                    visit(d);
                }
            }
            visit('$main');
            all = all.filter((v) => used.has(v.name));
        }

        // Sort functions
        let sorted = topologicalSort(all, (f) => Array.from(f.depends).map((v) => this.#functions.get(v)!!));

        // Render
        let res = '';
        for (let h of this.#headers) {
            if (res !== '') {
                res += '\n';
            }
            res += h;
        }
        for (let f of sorted) {
            if (res !== '') {
                res += '\n\n';
            }
            res += f.code;
        }
        return res;
    }

    //
    // Declaration
    //

    fun(name: string, handler: () => void) {
        if (this.#functions.has(name)) {
            throw new Error(`Function ${name} already defined`); // Should not happen
        }
        if (!!this.#pendingWriter || !!this.#pendingDepends) {
            throw new Error(`Nested function definition is not supported`); // Should not happen
        }

        // console.log(`Rendering: "${name}"`);

        // Write function
        this.#pendingWriter = new Writer();
        this.#pendingDepends = new Set();
        this.#pendingName = name;
        handler();
        let code = this.#pendingWriter.end();
        let depends = this.#pendingDepends;
        this.#pendingDepends = null;
        this.#pendingWriter = null;
        this.#pendingName = null;
        this.#functions.set(name, { name, code, depends });
    }

    used(name: string) {
        if (this.#pendingName !== name) {
            this.#pendingDepends!!.add(name);
        }
        return name;
    }

    currentContext() {
        return this.#pendingName;
    }

    //
    // Writers
    //

    inIndent = (handler: () => void) => {
        this.#pendingWriter!.inIndent(handler);
    };

    append(src: string = '') {
        this.#pendingWriter!.append(src);
    }

    write(src: string = '') {
        this.#pendingWriter!.write(src);
    }

    debug(id?: number | undefined | null | string) {
        if (enabledDebug(this.ctx)) {
            if (typeof id === 'string') {
                this.used('__tact_debug_str');
                this.append(`__tact_debug_str("${id}");`);
            } else {
                this.used('__tact_debug');
                let v = (id === undefined || id === null) ? (this.#nextId++) : id;
                this.append(`__tact_debug(${v});`);
            }
        }
    }

    //
    // Headers
    //

    header(src: string) {
        this.#headers.push(src);
    }
}