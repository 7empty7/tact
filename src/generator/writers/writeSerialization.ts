import { contractErrors } from "../../abi/errors";
import { StorageAllocation, StorageCell, StorageField } from "../../storage/StorageAllocation";
import { TypeDescription } from "../../types/types";
import { WriterContext } from "../Writer";
import { resolveFuncType } from "./resolveFuncType";
import { resolveFuncTypeUnpack } from "./resolveFuncTypeUnpack";

//
// Serializer
//

function writeSerializerField(f: StorageField, index: number, ctx: WriterContext, optional: boolean = false) {

    // Handle optional

    if (f.kind === 'optional') {
        ctx.append(`if (null?(v'${f.name})) {`);
        ctx.inIndent(() => {
            ctx.append(`build_${index} = store_int(build_${index}, false, 1);`);
        });
        ctx.append(`} else {`);
        ctx.inIndent(() => {
            ctx.append(`build_${index} = store_int(build_${index}, true, 1);`);
            writeSerializerField(f.inner, index, ctx, true);
        });
        ctx.append(`}`);
        return;
    }

    // Handle primitives

    if (f.kind === 'int') {
        ctx.append(`build_${index} = store_int(build_${index}, v'${f.name}, ${f.bits});`);
        return;
    }

    if (f.kind === 'uint') {
        ctx.append(`build_${index} = store_uint(build_${index}, v'${f.name}, ${f.bits});`);
        return;
    }

    if (f.kind === 'coins') {
        ctx.append(`build_${index} = store_coins(build_${index}, v'${f.name});`);
        return;
    }

    if (f.kind === 'slice') {
        ctx.append(`build_${index} = store_ref(build_${index}, v'${f.name}.end_cell());`);
        return;
    }

    if (f.kind === 'cell') {
        ctx.append(`build_${index} = store_ref(build_${index}, v'${f.name});`);
        return;
    }

    if (f.kind === 'address') {
        ctx.used(`__tact_store_address`);
        ctx.append(`build_${index} = __tact_store_address(build_${index}, v'${f.name});`);
        return;
    }

    if (f.kind === 'map') {
        ctx.append(`build_${index} = store_dict(build_${index}, v'${f.name});`);
        return;
    }

    if (f.kind === 'remaining') {
        ctx.append(`build_${index} = store_slice(build_${index}, v'${f.name});`);
        return;
    }

    if (f.kind === 'bytes') {
        ctx.append(`build_${index} = store_slice(build_${index}, v'${f.name});`);
        return;
    }

    // Handle structs

    if (f.kind === 'struct') {
        if (optional) {
            ctx.used(`__gen_write_${f.type.name}`);
            ctx.used(`__gen_${f.type.name}_not_null`);
            ctx.append(`build_${index} = __gen_write_${f.type.name}(build_${index}, __gen_${f.type.name}_not_null(v'${f.name}));`);
        } else {
            ctx.used(`__gen_write_${f.type.name}`);
            ctx.append(`build_${index} = __gen_write_${f.type.name}(build_${index}, ${resolveFuncTypeUnpack(f.type, `v'${f.name}`, ctx)});`);
        }
        return;
    }

    throw Error('Unsupported field kind: ' + f.kind);
}

function writeSerializerCell(cell: StorageCell, index: number, ctx: WriterContext) {

    // Write fields
    for (let f of cell.fields) {
        writeSerializerField(f, index, ctx);
    }

    // Tail
    if (cell.next) {
        ctx.append(`var build_${index + 1} = begin_cell();`);
        writeSerializerCell(cell.next!, index + 1, ctx);
        ctx.append(`build_${index} = store_ref(build_${index}, build_${index + 1}.end_cell());`);
    }
}

export function writeSerializer(name: string, allocation: StorageAllocation, ctx: WriterContext) {

    // Write to builder
    ctx.fun(`__gen_write_${name}`, () => {
        ctx.append(`builder __gen_write_${name}(builder build_0, ${resolveFuncType(allocation.type, ctx)} v) inline {`);
        ctx.inIndent(() => {
            ctx.append(`var ${resolveFuncTypeUnpack(allocation.type, `v`, ctx)} = v;`)
            if (allocation.prefix) {
                ctx.append(`build_0 = store_uint(build_0, ${allocation.prefix}, 32);`);
            }
            if (allocation.fields.length > 0) {
                writeSerializerCell(allocation.root, 0, ctx);
            }
            ctx.append(`return build_0;`);
        });
        ctx.append(`}`);
    });

    // Write to cell
    ctx.fun(`__gen_writecell_${name}`, () => {
        ctx.append(`cell __gen_writecell_${name}(${resolveFuncType(allocation.type, ctx)} v) inline {`);
        ctx.inIndent(() => {
            ctx.used(`__gen_write_${name}`);
            ctx.append(`return __gen_write_${name}(begin_cell(), v).end_cell();`);
        });
        ctx.append(`}`);
    });

    // Write to slice
    ctx.fun(` __gen_writeslice_${name}`, () => {
        ctx.append(`slice __gen_writeslice_${name}(${resolveFuncType(allocation.type, ctx)} v}) inline {`);
        ctx.inIndent(() => {
            ctx.used(`__gen_writecell_${name}`);
            ctx.append(`return __gen_writecell_${name}(v).begin_parse();`);
        });
        ctx.append(`}`);
    });

    // Write to opt cell
    ctx.fun(`__gen_writecellopt_${name}`, () => {
        ctx.append(`cell __gen_writecellopt_${name}(tuple v) inline {`);
        ctx.inIndent(() => {

            // If null
            ctx.append(`if (null?(v)) {`);
            ctx.inIndent(() => {
                ctx.append(`return null();`);
            });
            ctx.append(`}`);

            // If not null
            ctx.used(`__gen_writecell_${name}`);
            ctx.append(`return __gen_writecell_${name}(v);`);
        });
        ctx.append(`}`);
    });
}

//
// Parser
//

function writeFieldParser(f: StorageField, index: number, ctx: WriterContext, optional: boolean = false) {
    let varName = optional ? `v'${f.name}` : `var v'${f.name}`;

    // Handle optional

    if (f.kind === 'optional') {
        ctx.append(`${varName} = null();`);
        ctx.append(`if (sc_${index}~load_int(1)) {`);
        ctx.inIndent(() => {
            writeFieldParser(f.inner, index, ctx, true);
        });
        ctx.append('}');
        return;
    }

    // Handle primitive values

    if (f.kind === 'int') {
        ctx.append(`${varName} = sc_${index}~load_int(${f.bits});`);
        return;
    }

    if (f.kind === 'uint') {
        ctx.append(`${varName} = sc_${index}~load_uint(${f.bits});`);
        return;
    }

    if (f.kind === 'coins') {
        ctx.append(`${varName} = sc_${index}~load_coins();`);
        return;
    }

    if (f.kind === 'slice') {
        ctx.append(`${varName} = sc_${index}~load_ref().begin_parse();`);
        return;
    }

    if (f.kind === 'cell') {
        ctx.append(`${varName} = sc_${index}~load_ref();`);
        return;
    }

    if (f.kind === 'address') {
        ctx.used(`__tact_load_address`);
        ctx.append(`${varName} = sc_${index}~__tact_load_address();`);
        return;
    }

    if (f.kind === 'map') {
        ctx.append(`${varName} = sc_${index}~load_dict();`);
        return;
    }

    if (f.kind === 'remaining') {
        ctx.append(`${varName} = sc_${index};`);
        return;
    }

    if (f.kind === 'bytes') {
        ctx.append(`${varName} = sc_${index}~load_bits(${f.bytes * 8});`);
        return;
    }

    // Handle structs

    if (f.kind === 'struct') {
        if (optional) {
            ctx.used(`__gen_read_${f.type.name}`);
            ctx.used(`__gen_${f.type.name}_as_optional`);
            ctx.append(`${varName} = __gen_${f.type.name}_as_optional(sc_${index}~__gen_read_${f.type.name}());`);
        } else {
            ctx.used(`__gen_read_${f.type.name}`);
            ctx.append(`${varName} = sc_${index}~__gen_read_${f.type.name}();`);
        }
        return;
    }

    throw Error('Unsupported field kind: ' + f.kind);
}

function writeCellParser(cell: StorageCell, index: number, ctx: WriterContext): number {

    // Write current fields
    for (let f of cell.fields) {
        writeFieldParser(f, index, ctx);
    }

    // Handle next cell
    if (cell.next) {
        ctx.append(`slice sc_${index + 1} = preload_ref(sc_${index}).begin_parse();`);
        return writeCellParser(cell.next, index + 1, ctx);
    } else {
        return index;
    }
}

export function writeParser(name: string, allocation: StorageAllocation, ctx: WriterContext) {
    ctx.fun(`__gen_read_${name}`, () => {
        ctx.append(`(slice, (${resolveFuncType(allocation.type, ctx)})) __gen_read_${name}(slice sc_0) inline {`);
        ctx.inIndent(() => {

            // Check prefix
            if (allocation.prefix) {
                ctx.append(`throw_unless(${contractErrors.invalidPrefix.id}, sc_0~load_uint(32) == ${allocation.prefix});`);
            }

            // Write cell parser
            let out = writeCellParser(allocation.root, 0, ctx);

            // Compile tuple
            ctx.append(`return (sc_${out}, (${allocation.fields.map((v) => `v'${v.name}`).join(', ')}));`);
        });
        ctx.append("}");
    });

    ctx.fun(`__gen_readopt_${name}`, () => {
        ctx.append(`tuple __gen_readopt_${name}(cell cl) inline {`);
        ctx.inIndent(() => {
            ctx.debug(`Invoke __gen_readopt_${name}`);

            // Handle null
            ctx.append(`if (null?(cl)) {`);
            ctx.inIndent(() => {
                ctx.append(`return null();`);
            });
            ctx.append(`}`);
            ctx.append(`var sc_0 = cl.begin_parse();`);

            // Check prefix
            if (allocation.prefix) {
                ctx.append(`throw_unless(${contractErrors.invalidPrefix.id}, sc_0~load_uint(32) == ${allocation.prefix});`);
            }

            // Write cell parser
            writeCellParser(allocation.root, 0, ctx);

            // Compile tuple
            ctx.used(`__tact_tuple_create_${allocation.fields.length}`);
            ctx.append(`return __tact_tuple_create_${allocation.fields.length}(${allocation.fields.map((v) => `v'${v.name}`).join(', ')});`);
        });
        ctx.append("}");
    });
}

//
// Storage
//

export function writeStorageOps(type: TypeDescription, ctx: WriterContext) {

    // Load function
    ctx.fun(`__gen_load_${type.name}`, () => {
        ctx.append(`${resolveFuncType(type, ctx)} __gen_load_${type.name}() inline {`); // NOTE: Inline function
        ctx.inIndent(() => {

            // Load data slice
            ctx.append(`slice sc = get_data().begin_parse();`);

            // Load context
            ctx.used(`__tact_context`);
            ctx.append(`__tact_context_sys = sc~load_ref();`);

            // Load data
            ctx.used(`__gen_read_${type.name}`);
            ctx.append(`return sc~__gen_read_${type.name}();`);
        });
        ctx.append(`}`);
    });

    // Store function
    ctx.fun(`__gen_store_${type.name}`, () => {
        ctx.append(`() __gen_store_${type.name}(${resolveFuncType(type, ctx)} v) impure inline {`); // NOTE: Impure function
        ctx.inIndent(() => {
            ctx.append(`builder b = begin_cell();`);

            // Persist system cell
            ctx.used(`__tact_context`);
            ctx.append(`b = b.store_ref(__tact_context_sys);`);

            // Build data
            ctx.used(`__gen_write_${type.name}`);
            ctx.append(`b = __gen_write_${type.name}(b, v);`);

            // Persist data
            ctx.append(`set_data(b.end_cell());`);
        });
        ctx.append(`}`);
    });
}