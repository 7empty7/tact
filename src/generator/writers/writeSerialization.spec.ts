import { CompilerContext } from "../../ast/context";
import { getAllocation, resolveAllocations } from "../../storage/resolveAllocation";
import { getType, resolveTypeDescriptors } from "../../types/resolveTypeDescriptors";
import { Writer, WriterContext } from "../Writer";
import { writeParser, writeSerializer } from "./writeSerialization";

const code = `
primitive Int;
primitive Bool;
primitive Builder;
primitive Cell;
primitive Slice;

struct A {
    a: Int;
    b: Int;
    c: Int?;
    d: Bool;
    e: Bool?;
    f: Int;
    g: Int;
}

struct B {
    a: Int;
    b: Int;
    c: Int?;
    d: Bool;
    e: Bool?;
    f: Int;
    g: Int;
}

struct C {
    a: Cell;
    b: Cell?;
    c: Slice?;
    d: Slice?;
    e: Bool;
    f: Int;
    g: Int;
}
`;

describe('writeSerialization', () => {
    for (let s of ['A', 'B', 'C']) {
        it('should write serializer for ' + s, () => {
            let ctx = CompilerContext.fromSources([code]);
            ctx = resolveTypeDescriptors(ctx);
            ctx = resolveAllocations(ctx);
            let w = new Writer();
            let wctx = new WriterContext();
            writeSerializer(ctx, s, getAllocation(ctx, s), w, wctx);
            writeParser(ctx, s, getAllocation(ctx, s), w, wctx);
            expect(w.end().toString()).toMatchSnapshot();
        });
    }
});