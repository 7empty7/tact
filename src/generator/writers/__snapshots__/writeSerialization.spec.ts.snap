// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeSerialization should write serializer for A 1`] = `
"forall X -> tuple __tact_set(tuple x, X v, int i) asm "SETINDEXVARQ";

forall X -> X __tact_not_null(X x) { throw_if(14, null?(x)); return x; }

(slice, slice) __tact_load_address(slice cs) inline {
    slice raw = cs~load_msg_addr();
    return (cs, raw);
}

builder __tact_store_address(builder b, slice address) inline {
    b = b.store_slice(address);
    return b;
}

slice __tact_compute_contract_address(int chain, cell code, cell data) {
    var b = begin_cell();
    b = b.store_uint(0, 2);
    b = b.store_uint(3, 2);
    b = b.store_uint(0, 1);
    b = b.store_ref(code);
    b = b.store_ref(data);
    var hash = cell_hash(b.end_cell());
    var b2 = begin_cell();
    b2 = b2.store_uint(2, 2);
    b2 = b2.store_uint(0, 1);
    b2 = b2.store_int(chain, 8);
    b2 = b2.store_uint(hash, 256);
    return b2.end_cell().begin_parse();
}

forall X -> tuple __tact_to_tuple(X x) impure asm "NOP";

forall X -> X __tact_from_tuple(tuple x) impure asm "NOP";

(cell, ()) __tact_dict_set_int_cell(cell d, int kl, int k, cell v) {
    if (null?(v)) {
        var (r, ok) = idict_delete?(d, kl, k);
        return (r, ());
    } else {
        return (idict_set_ref(d, kl, k, v), ());
    }
}

(cell, ()) __tact_dict_set_int_int(cell d, int kl, int k, int v, int vl) {
    if (null?(v)) {
        var (r, ok) = idict_delete?(d, kl, k);
        return (r, ());
    } else {
        return (idict_set_ref(d, kl, k, begin_cell().store_int(v, vl).end_cell()), ());
    }
}

builder __gen_write_A(builder build_0, int v'a, int v'b, int v'c, int v'd, int v'e, int v'f, int v'g) inline {
    build_0 = store_int(build_0, v'a, 257);
    build_0 = store_int(build_0, v'b, 257);
    if (null?(v'c)) {
        build_0 = store_int(build_0, false, 1);
    } else {
        build_0 = store_int(build_0, true, 1);
        build_0 = store_int(build_0, v'c, 257);
    }
    build_0 = store_int(build_0, v'd, 1);
    if (null?(v'e)) {
        build_0 = store_int(build_0, false, 1);
    } else {
        build_0 = store_int(build_0, true, 1);
        build_0 = store_int(build_0, v'e, 1);
    }
    var build_1 = begin_cell();
    build_1 = store_int(build_1, v'f, 257);
    build_1 = store_int(build_1, v'g, 257);
    build_0 = store_ref(build_0, build_1.end_cell());
    return build_0;
}

cell __gen_writecell_A(int v'a,int v'b,int v'c,int v'd,int v'e,int v'f,int v'g) inline {
    return __gen_write_A(begin_cell(), v'a, v'b, v'c, v'd, v'e, v'f, v'g).end_cell();
}

slice __gen_writeslice_A(int v'a,int v'b,int v'c,int v'd,int v'e,int v'f,int v'g) inline {
    return __gen_writecell_A(v'a, v'b, v'c, v'd, v'e, v'f, v'g).begin_parse();
}

(slice, (int, int, int, int, int, int, int)) __gen_read_A(slice sc) inline {
    var a = sc~load_int(257);
    var b = sc~load_int(257);
    var c = null();
    if (sc~load_int(1)) {
        var c = sc~load_int(257);
    }
    var d = sc~load_int(1);
    var e = null();
    if (sc~load_int(1)) {
        var e = sc~load_int(1);
    }
    sc = (sc~load_ref()).begin_parse();
    var f = sc~load_int(257);
    var g = sc~load_int(257);
    return (sc, (a, b, c, d, e, f, g));
}"
`;

exports[`writeSerialization should write serializer for B 1`] = `
"forall X -> tuple __tact_set(tuple x, X v, int i) asm "SETINDEXVARQ";

forall X -> X __tact_not_null(X x) { throw_if(14, null?(x)); return x; }

(slice, slice) __tact_load_address(slice cs) inline {
    slice raw = cs~load_msg_addr();
    return (cs, raw);
}

builder __tact_store_address(builder b, slice address) inline {
    b = b.store_slice(address);
    return b;
}

slice __tact_compute_contract_address(int chain, cell code, cell data) {
    var b = begin_cell();
    b = b.store_uint(0, 2);
    b = b.store_uint(3, 2);
    b = b.store_uint(0, 1);
    b = b.store_ref(code);
    b = b.store_ref(data);
    var hash = cell_hash(b.end_cell());
    var b2 = begin_cell();
    b2 = b2.store_uint(2, 2);
    b2 = b2.store_uint(0, 1);
    b2 = b2.store_int(chain, 8);
    b2 = b2.store_uint(hash, 256);
    return b2.end_cell().begin_parse();
}

forall X -> tuple __tact_to_tuple(X x) impure asm "NOP";

forall X -> X __tact_from_tuple(tuple x) impure asm "NOP";

(cell, ()) __tact_dict_set_int_cell(cell d, int kl, int k, cell v) {
    if (null?(v)) {
        var (r, ok) = idict_delete?(d, kl, k);
        return (r, ());
    } else {
        return (idict_set_ref(d, kl, k, v), ());
    }
}

(cell, ()) __tact_dict_set_int_int(cell d, int kl, int k, int v, int vl) {
    if (null?(v)) {
        var (r, ok) = idict_delete?(d, kl, k);
        return (r, ());
    } else {
        return (idict_set_ref(d, kl, k, begin_cell().store_int(v, vl).end_cell()), ());
    }
}

builder __gen_write_B(builder build_0, int v'a, int v'b, int v'c, int v'd, int v'e, int v'f, int v'g) inline {
    build_0 = store_int(build_0, v'a, 257);
    build_0 = store_int(build_0, v'b, 257);
    if (null?(v'c)) {
        build_0 = store_int(build_0, false, 1);
    } else {
        build_0 = store_int(build_0, true, 1);
        build_0 = store_int(build_0, v'c, 257);
    }
    build_0 = store_int(build_0, v'd, 1);
    if (null?(v'e)) {
        build_0 = store_int(build_0, false, 1);
    } else {
        build_0 = store_int(build_0, true, 1);
        build_0 = store_int(build_0, v'e, 1);
    }
    var build_1 = begin_cell();
    build_1 = store_int(build_1, v'f, 257);
    build_1 = store_int(build_1, v'g, 257);
    build_0 = store_ref(build_0, build_1.end_cell());
    return build_0;
}

cell __gen_writecell_B(int v'a,int v'b,int v'c,int v'd,int v'e,int v'f,int v'g) inline {
    return __gen_write_B(begin_cell(), v'a, v'b, v'c, v'd, v'e, v'f, v'g).end_cell();
}

slice __gen_writeslice_B(int v'a,int v'b,int v'c,int v'd,int v'e,int v'f,int v'g) inline {
    return __gen_writecell_B(v'a, v'b, v'c, v'd, v'e, v'f, v'g).begin_parse();
}

(slice, (int, int, int, int, int, int, int)) __gen_read_B(slice sc) inline {
    var a = sc~load_int(257);
    var b = sc~load_int(257);
    var c = null();
    if (sc~load_int(1)) {
        var c = sc~load_int(257);
    }
    var d = sc~load_int(1);
    var e = null();
    if (sc~load_int(1)) {
        var e = sc~load_int(1);
    }
    sc = (sc~load_ref()).begin_parse();
    var f = sc~load_int(257);
    var g = sc~load_int(257);
    return (sc, (a, b, c, d, e, f, g));
}"
`;

exports[`writeSerialization should write serializer for C 1`] = `
"forall X -> tuple __tact_set(tuple x, X v, int i) asm "SETINDEXVARQ";

forall X -> X __tact_not_null(X x) { throw_if(14, null?(x)); return x; }

(slice, slice) __tact_load_address(slice cs) inline {
    slice raw = cs~load_msg_addr();
    return (cs, raw);
}

builder __tact_store_address(builder b, slice address) inline {
    b = b.store_slice(address);
    return b;
}

slice __tact_compute_contract_address(int chain, cell code, cell data) {
    var b = begin_cell();
    b = b.store_uint(0, 2);
    b = b.store_uint(3, 2);
    b = b.store_uint(0, 1);
    b = b.store_ref(code);
    b = b.store_ref(data);
    var hash = cell_hash(b.end_cell());
    var b2 = begin_cell();
    b2 = b2.store_uint(2, 2);
    b2 = b2.store_uint(0, 1);
    b2 = b2.store_int(chain, 8);
    b2 = b2.store_uint(hash, 256);
    return b2.end_cell().begin_parse();
}

forall X -> tuple __tact_to_tuple(X x) impure asm "NOP";

forall X -> X __tact_from_tuple(tuple x) impure asm "NOP";

(cell, ()) __tact_dict_set_int_cell(cell d, int kl, int k, cell v) {
    if (null?(v)) {
        var (r, ok) = idict_delete?(d, kl, k);
        return (r, ());
    } else {
        return (idict_set_ref(d, kl, k, v), ());
    }
}

(cell, ()) __tact_dict_set_int_int(cell d, int kl, int k, int v, int vl) {
    if (null?(v)) {
        var (r, ok) = idict_delete?(d, kl, k);
        return (r, ());
    } else {
        return (idict_set_ref(d, kl, k, begin_cell().store_int(v, vl).end_cell()), ());
    }
}

builder __gen_write_C(builder build_0, cell v'a, cell v'b, slice v'c, slice v'd, int v'e, int v'f, int v'g, slice v'h) inline {
    build_0 = store_ref(build_0, v'a);
    if (null?(v'b)) {
        build_0 = store_int(build_0, false, 1);
    } else {
        build_0 = store_int(build_0, true, 1);
        build_0 = store_ref(build_0, v'b);
    }
    if (null?(v'c)) {
        build_0 = store_int(build_0, false, 1);
    } else {
        build_0 = store_int(build_0, true, 1);
        build_0 = store_ref(build_0, v'c.end_cell());
    }
    var build_1 = begin_cell();
    if (null?(v'd)) {
        build_1 = store_int(build_1, false, 1);
    } else {
        build_1 = store_int(build_1, true, 1);
        build_1 = store_ref(build_1, v'd.end_cell());
    }
    build_1 = store_int(build_1, v'e, 1);
    build_1 = store_int(build_1, v'f, 257);
    build_1 = store_int(build_1, v'g, 257);
    build_1 = __tact_store_address(build_1, v'h);
    build_0 = store_ref(build_0, build_1.end_cell());
    return build_0;
}

cell __gen_writecell_C(cell v'a,cell v'b,slice v'c,slice v'd,int v'e,int v'f,int v'g,slice v'h) inline {
    return __gen_write_C(begin_cell(), v'a, v'b, v'c, v'd, v'e, v'f, v'g, v'h).end_cell();
}

slice __gen_writeslice_C(cell v'a,cell v'b,slice v'c,slice v'd,int v'e,int v'f,int v'g,slice v'h) inline {
    return __gen_writecell_C(v'a, v'b, v'c, v'd, v'e, v'f, v'g, v'h).begin_parse();
}

(slice, (cell, cell, slice, slice, int, int, int, slice)) __gen_read_C(slice sc) inline {
    var a = sc~load_ref();
    var b = null();
    if (sc~load_int(1)) {
        var b = sc~load_ref();
    }
    var c = null();
    if (sc~load_int(1)) {
        var c = sc~load_ref().begin_parse();
    }
    sc = (sc~load_ref()).begin_parse();
    var d = null();
    if (sc~load_int(1)) {
        var d = sc~load_ref().begin_parse();
    }
    var e = sc~load_int(1);
    var f = sc~load_int(257);
    var g = sc~load_int(257);
    var h = sc~__tact_load_address();
    return (sc, (a, b, c, d, e, f, g, h));
}"
`;
